collector/nvml/status.go

```diff
@@ -0,0 +1,115 @@
+// +build !darwin
+// +build !debug
+
+package nvml
+
+/*
+
+#cgo CFLAGS: -std=c99 -I .
+#cgo LDFLAGS: -L${SRCDIR}/ -lnvidia-ml -lpthread -ldl -lrt
+#include "nvidia-smi.h"
+#include <stdlib.h>
+
+int gpu_stats_ok(int *pCount, GPU_Stats **ppStats) {
+    GPU_Stats * pBuf = (GPU_Stats *)calloc(1, sizeof(GPU_Stats));
+    if (pBuf == NULL) {
+        return -1;
+    }
+    *pCount = 1;
+    *ppStats = pBuf;
+    return 0;
+}
+
+int gpu_stats_ok_throttle(int *pCount, GPU_Stats **ppStats) {
+    GPU_Stats * pBuf = (GPU_Stats *)calloc(1, sizeof(GPU_Stats));
+    if (pBuf == NULL) {
+        return -1;
+    }
+    *pCount = 1;
+    pBuf->throttle = 0x8000000000000000LL;
+    *ppStats = pBuf;
+    return 0;
+}
+
+int gpu_stats_failed(int *pCount, GPU_Stats **ppStats) {
+    return -1;
+}
+
+*/
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type Nvidia_GPU_Stats struct {
+	ClockMem      uint // GPU memory clock in Mhz
+	ClockGraphics uint // GPU graphics clock in Mhz
+	Throttle      uint // throttle reason
+	PerfState     uint // performance state    C.uint 0: max / 15: min
+	Temperature   uint // GPU temperature in Celsius degrees
+	UtilGPU       uint // percentage of time during kernels are executing on the GPU.
+	UtilMem       uint // percentage of time during memory is being read or written.
+	MemUsage      uint // percentage of used memory size
+	ID            uint // device ID
+}
+
+var fn_get_stats = get_stats_real
+
+// get_stats_real calls real function to get GPU stats
+func get_stats_real(cnt *C.int, cs **C.GPU_Stats) C.int {
+	return C.get_gpu_stats(cnt, cs)
+}
+
+// GetGPUStats() returns stats of each GPU on host
+func GetGPUStats() []Nvidia_GPU_Stats {
+	var stats []Nvidia_GPU_Stats
+	var cs *C.GPU_Stats
+	var cnt C.int
+	if fn_get_stats(&cnt, &cs) == -1 {
+		return nil
+	}
+	defer C.free(unsafe.Pointer(cs))
+
+	// convert C array to slice
+	slice := (*[1 << 30]C.GPU_Stats)(unsafe.Pointer(cs))[:cnt:cnt]
+
+	stats = make([]Nvidia_GPU_Stats, cnt)
+	for i, val := range slice {
+		stats[i].ClockMem = uint(val.clock_mem)
+		stats[i].ClockGraphics = uint(val.clock_graphics)
+		if val.throttle >= 0xFFFF {
+			// the larget value of nvmlClocksThrottleReasonUnknown 0x8000000000000000LL
```

Comment:
Have you taken this code from elsewhere? Copied code needs to be properly attributed.

Reply:
Thanks @brian-brazil .
The author of this piece of code is from my colleague. I've had his authorization.